### PR TITLE
Add BITAP algorithm for search with SQL integration

### PR DESCRIPTION
**Pull request by shubhraneel**
Written the Bitap function in C. Not yet integrated or tested with MySQL but tested it with words input by the user.

### CODE REVIEW

**Comment by shubhraneel** [comment]
This PR if merged will introduce the following changes

- The bitap function implemented along with SQL integration
- The Makefile for building the `db` container shall be executed automatically again (if testing the function Bitap we need to run `install.sh` everytime a change is made to Bitap, including this time)
- The bitap.c file previously included is deleted as bitap.cpp in mariadb-udf-docker is used instead
- The docker yaml template is changed to automatically start the container for image `mariadb-udf` instead of `mariadb:latest`

**Comment by grapheo12** [comment]
This is some excellent work! I'll review it tonight.

File: docker-compose.yml.template [hunk 1]
```diff
@@ -16,7 +16,7 @@ services:
     depends_on:
       - web
   db:
-    image: mariadb:latest
+    image: mariadb-udf
     restart: always
```
Comment (grapheo12) [inline docker-compose.yml.template:20]: Add a `build: ./mariadb-udf-docker` here (as all other images have)

File: install.sh [hunk 2]
```diff
@@ -37,6 +37,10 @@ else
     exit
 fi
 
+echo "Building mariadb-udf image"
+
+make -C ./mariadb-udf-docker/
```
Comment (grapheo12) [inline install.sh:42]: You need to check if `make` and `g++` is available in the system. This type of check is done Line 4. Edit that.

File: mariadb-udf-docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,11 @@
+FROM mariadb:latest
+
+ENV MYSQL_DATABASE='dummy'
+ENV MYSQL_USER='admin'
+ENV MYSQL_PASSWORD='abc123'
+ENV MYSQL_ROOT_PASSWORD='aabbcc112233'
+ENV MYSQL_ROOT_HOST='%'
```
Comment (grapheo12) [inline mariadb-udf-docker/Dockerfile:7]: Remove the `ENV` lines. We provide them from the docker-compose.yml.

File: mariadb-udf-docker/README.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+To change the function, modify `mariadb.hpp` and `bitap.cpp`.
+You may also rename `bitap.cpp` to your liking. Then you also need to modify `Makefile`.
+
+To build the image, run `make`.
+
+An image by the name `mariadb-udf` will be created.
+This image will load the user-defined function on its startup.
+
+For a comprehensive tutorial on how to create a user-defined function, follow:
+[https://www.oreilly.com/library/view/mysql-reference-manual/0596002653/ch09s02.html](https://www.oreilly.com/library/view/mysql-reference-manual/0596002653/ch09s02.html)
```
Comment (grapheo12) [inline mariadb-udf-docker/README.md:16]: Delete these lines. Just write down the source repo.

**Review by grapheo12 — CHANGES_REQUESTED** [review]
Please make the changes and thoroughly test, then commit.

**Comment by grapheo12** [comment]
@shubhraneel Please work on these changes. You are doing great.

**Comment by shubhraneel** [comment]
I have a doubt. Since I have learned Docker recently, I may be wrong. If I use `build: ./mariadb-udf-docker`, it builds a new image and then a container but the Makefile in `mariadb-udf-docker` already creates an image named `mariadb-udf` so I thought I should use that image created by Makefile itself. 

Also, in case you are having a problem where the function does not work. I thing deleting the database or using a new folder for the database while running install.sh works.

**Comment by grapheo12** [comment]
@shubhraneel I believe the best course of action will be not to build the image in the `install.sh` script. Just run the makefile locally and commit `bitap.so`. In this way, you won't require to compile it in user's computer and the build can happen directly.
It is better to keep the whole stack in a single docker-compose file. Otherwise all images except the db one will be properly tagged with `iqps_` prefix.

**Comment by grapheo12** [comment]
One other way is to (and I believe it is better) is to just compile the .so file using the `make` call in the `install.sh` file: `make bitap.so -C mariadb-udf-docker`

**Comment by shubhraneel** [comment]
I removed the `build` command from the Makefile. I tested it and it is working. That way the redundant image is not created.

**Comment by grapheo12** [comment]
Did you remove the old images?

**Comment by grapheo12** [comment]
The PR looks good to me! Merging!

**Pull request merged**